### PR TITLE
[LYN-14352] Asset Manager - Fix deadlock issue caused by calling GetAsset inside AssetBus callback

### LINKED ISSUE #11014: Editor lockups when entering game mode, 50% repro, deadlocked loading a material
**Describe the bug**
In a level with an EMFX actor, entering game mode occasionally locks up due to a deadlock.

```
 	[Inline Frame] AtomLyIntegration_CommonFeatures.Editor.dll!AZStd::recursive_mutex::lock() Line 77	C++
 	[Inline Frame] AtomLyIntegration_CommonFeatures.Editor.dll!AZStd::scoped_lock<AZStd::recursive_mutex>::{ctor}(AZStd::recursive_mutex &) Line 46	C++
>	AtomLyIntegration_CommonFeatures.Editor.dll!AZ::Data::AssetManager::GetAssetContainer(AZ::Data::Asset<AZ::Data::AssetData> asset, const AZ::Data::AssetLoadParameters & loadParams, bool isReload) Line 2312	C++
 	AtomLyIntegration_CommonFeatures.Editor.dll!AZ::Data::AssetManager::GetAsset(const AZ::Data::AssetId & assetId, const AZ::Uuid & assetType, AZ::Data::AssetLoadBehaviorNamespace::AssetLoadBehavior assetReferenceLoadBehavior, const AZ::Data::AssetLoadParameters & loadParams) Line 931	C++
 	AtomLyIntegration_CommonFeatures.Editor.dll!AZ::Data::AssetInternal::GetAsset(const AZ::Data::AssetId & id, const AZ::Uuid & type, AZ::Data::AssetLoadBehaviorNamespace::AssetLoadBehavior assetReferenceLoadBehavior, const AZ::Data::AssetLoadParameters & loadParams) Line 113	C++
 	AtomLyIntegration_CommonFeatures.Editor.dll!AZ::Data::Asset<AZ::RPI::MaterialAsset>::QueueLoad(const AZ::Data::AssetLoadParameters & loadParams) Line 1132	C++
 	AtomLyIntegration_CommonFeatures.Editor.dll!AZ::Render::MaterialComponentController::LoadMaterials() Line 232	C++
 	AtomLyIntegration_CommonFeatures.Editor.dll!AZ::Render::MaterialComponentController::OnMaterialAssignmentSlotsChanged() Line 759	C++
 	[Inline Frame] EMotionFX_Atom.Editor.dll!AZStd::Internal::INVOKE(void(AZ::Render::MaterialReceiverNotifications::*)() &) Line 177	C++
 	[Inline Frame] EMotionFX_Atom.Editor.dll!AZStd::invoke(void(AZ::Render::MaterialReceiverNotifications::*)() &) Line 54	C++
 	[Inline Frame] EMotionFX_Atom.Editor.dll!AZ::EBusEventProcessingPolicy::Call(void(AZ::Render::MaterialReceiverNotifications::*)() &) Line 437	C++
 	EMotionFX_Atom.Editor.dll!AZ::Internal::EBusContainer<AZ::Render::MaterialReceiverNotifications,AZ::Render::MaterialReceiverNotifications,1,1>::Dispatcher<AZ::EBus<AZ::Render::MaterialReceiverNotifications,AZ::Render::MaterialReceiverNotifications>>::Event<void (__cdecl AZ::Render::MaterialReceiverNotifications::*)(void)>(const AZ::EntityId & id, void(AZ::Render::MaterialReceiverNotifications::*)() && func) Line 146	C++
 	EMotionFX_Atom.Editor.dll!AZ::Render::AtomActorInstance::CreateSkinnedMeshInstance() Line 672	C++
 	EMotionFX_Atom.Editor.dll!AZ::Render::AtomActorInstance::Create() Line 462	C++
 	EMotionFX_Atom.Editor.dll!AZ::Render::AtomActorInstance::AtomActorInstance(AZ::EntityId entityId, const EMotionFX::Integration::EMotionFXPtr<EMotionFX::ActorInstance> & actorInstance, const AZ::Data::Asset<EMotionFX::Integration::ActorAsset> & asset, const AZ::Transform & worldTransform, EMotionFX::Integration::SkinningMethod skinningMethod) Line 61	C++
 	EMotionFX.Editor.dll!EMotionFX::Integration::EditorActorComponent::CheckActorCreation() Line 951	C++
 	EMotionFX.Editor.dll!EMotionFX::Integration::EditorActorComponent::OnAssetReady(AZ::Data::Asset<AZ::Data::AssetData> asset) Line 543	C++
 	[Inline Frame] EditorLib.dll!AZStd::Internal::INVOKE(void(AZ::Data::AssetEvents::*)(AZ::Data::Asset<AZ::Data::AssetData>) &) Line 177	C++
 	[Inline Frame] EditorLib.dll!AZStd::invoke(void(AZ::Data::AssetEvents::*)(AZ::Data::Asset<AZ::Data::AssetData>) &) Line 54	C++
 	[Inline Frame] EditorLib.dll!AZ::EBusEventProcessingPolicy::Call(void(AZ::Data::AssetEvents::*)(AZ::Data::Asset<AZ::Data::AssetData>) &) Line 437	C++
 	EditorLib.dll!AZ::Internal::EBusContainer<AZ::Data::AssetEvents,AZ::Data::AssetEvents,1,1>::Dispatcher<AZ::EBus<AZ::Data::AssetEvents,AZ::Data::AssetEvents>>::Event<void (__cdecl AZ::Data::AssetEvents::*)(AZ::Data::Asset<AZ::Data::AssetData>),AZ::Data::Asset<AZ::Data::AssetData> &>(const AZ::Data::AssetId & id, void(AZ::Data::AssetEvents::*)(AZ::Data::Asset<AZ::Data::AssetData>) && func, AZ::Data::Asset<AZ::Data::AssetData> & <args_0>) Line 146	C++
 	EditorLib.dll!AZ::Data::AssetManager::NotifyAssetReady(AZ::Data::Asset<AZ::Data::AssetData> asset) Line 1825	C++
 	[Inline Frame] EditorLib.dll!AZStd::Internal::INVOKE(void(AZ::Data::AssetManager::*)(AZ::Data::Asset<AZ::Data::AssetData>) &&) Line 177	C++
 	[Inline Frame] EditorLib.dll!AZStd::invoke(void(AZ::Data::AssetManager::*)(AZ::Data::Asset<AZ::Data::AssetData>) &&) Line 54	C++
 	[Inline Frame] EditorLib.dll!AZ::BusInternal::EBusBroadcastQueue<AZ::EBus<AZ::Data::AssetEvents,AZ::Data::AssetEvents>,AZ::BusInternal::EBusImplTraits<AZ::Data::AssetEvents,AZ::Data::AssetEvents>>::QueueFunction::__l5::<lambda_1>::operator()() Line 860	C++
 	[Inline Frame] EditorLib.dll!AZStd::Internal::INVOKE(AZ::BusInternal::EBusBroadcastQueue<AZ::EBus<AZ::Data::AssetEvents,AZ::Data::AssetEvents>,AZ::BusInternal::EBusImplTraits<AZ::Data::AssetEvents,AZ::Data::AssetEvents>>::QueueFunction::__l5::<lambda_1> &) Line 208	C++
 	[Inline Frame] EditorLib.dll!AZStd::invoke(AZ::BusInternal::EBusBroadcastQueue<AZ::EBus<AZ::Data::AssetEvents,AZ::Data::AssetEvents>,AZ::BusInternal::EBusImplTraits<AZ::Data::AssetEvents,AZ::Data::AssetEvents>>::QueueFunction::__l5::<lambda_1> &) Line 54	C++
 	[Inline Frame] EditorLib.dll!AZStd::Internal::function_util::invoke_void_return_wrapper<void>::call(AZ::BusInternal::EBusBroadcastQueue<AZ::EBus<AZ::Data::AssetEvents,AZ::Data::AssetEvents>,AZ::BusInternal::EBusImplTraits<AZ::Data::AssetEvents,AZ::Data::AssetEvents>>::QueueFunction::__l5::<lambda_1> &) Line 41	C++
 	EditorLib.dll!AZStd::Internal::function_util::get_invoker<void __cdecl(void),AZStd::Internal::function_util::function_obj_tag,AZ::Internal::EBusEnvironmentAllocator>::call<`AZ::BusInternal::EBusBroadcastQueue<AZ::EBus<AZ::Data::AssetEvents,AZ::Data::AssetEvents>,AZ::BusInternal::EBusImplTraits<AZ::Data::AssetEvents,AZ::Data::AssetEvents>>::QueueFunction<void (__cdecl AZ::Data::AssetManager::*)(AZ::Data::Asset<AZ::Data::AssetData>),AZ::Data::AssetManager *,AZ::Data::Asset<AZ::Data::AssetData>>'::`5'::<lambda_1>>(AZStd::Internal::function_util::function_buffer & function_obj_ptr) Line 148	C++
 	[Inline Frame] EditorLib.dll!AZStd::function_intermediate<void>::operator()() Line 604	C++
 	[Inline Frame] EditorLib.dll!AZStd::function<void __cdecl(void)>::operator()() Line 684	C++
 	EditorLib.dll!AZ::EBusQueuePolicy<1,AZ::EBus<AZ::Data::AssetEvents,AZ::Data::AssetEvents>,AZStd::recursive_mutex>::Execute() Line 267	C++
 	[Inline Frame] EditorLib.dll!AZ::BusInternal::EBusBroadcastQueue<AZ::EBus<AZ::Data::AssetEvents,AZ::Data::AssetEvents>,AZ::BusInternal::EBusImplTraits<AZ::Data::AssetEvents,AZ::Data::AssetEvents>>::ExecuteQueuedEvents() Line 366	C++
 	EditorLib.dll!AZ::Data::AssetManager::DispatchEvents() Line 590	C++
 	[Inline Frame] EditorLib.dll!AZStd::Internal::INVOKE(void(AZ::SystemTickEvents::*)() &) Line 177	C++
 	[Inline Frame] EditorLib.dll!AZStd::invoke(void(AZ::SystemTickEvents::*)() &) Line 54	C++
 	[Inline Frame] EditorLib.dll!AZ::EBusEventProcessingPolicy::Call(void(AZ::SystemTickEvents::*)() &) Line 437	C++
 	EditorLib.dll!AZ::Internal::EBusContainer<AZ::SystemTickEvents,AZ::SystemTickEvents,0,1>::Dispatcher<AZ::EBus<AZ::SystemTickEvents,AZ::SystemTickEvents>>::Broadcast<void (__cdecl AZ::SystemTickEvents::*)(void)>(void(AZ::SystemTickEvents::*)() && func) Line 1360	C++
 	EditorLib.dll!AZ::ComponentApplication::TickSystem() Line 1370	C++
 	EditorLib.dll!CCryEditApp::IdleProcessing(bool bBackgroundUpdate) Line 2279	C++
 	EditorLib.dll!CCryEditApp::OnIdle(long lCount) Line 2189	C++
 	EditorLib.dll!Editor::EditorQtApplication::maybeProcessIdle() Line 312	C++

```

**Assets required**
I was using this actor: https://www.mixamo.com/#/?page=1&query=jennifer&type=Character

**Steps to reproduce**
1. Open Editor
2. create or load a level with an actor
3. enter game mode a few times

**Expected behavior**
Game mode is entered

**Actual behavior**
Editor occasionally locks up.

**Found in Branch**
development

**Desktop/Device (please complete the following information):**
 - Device: PC
 - OS: Windows

### PR DESCRIPTION
**Pull request by amzn-mike**
## What does this PR do?

Added OnAsset[Ready/Reloaded/Error] events to the internal AssetLoadBus and switched AssetContainer over to using that ebus instead of AssetBus.

This fixes an issue where AssetContainer caused deadlocks with external systems trying to call GetAsset from within an AssetBus callback.

Fixes #11014 

## How was this PR tested?

Manually loading a level several times in the editor + running the added unit test

### CODE REVIEW

File: Code/Framework/AzCore/AzCore/Asset/AssetContainer.cpp [hunk 1]
```diff
@@ -32,7 +32,6 @@ namespace AZ::Data
                 "end up in a perpetual loading state if there is no top-level container signalling the completion of the full load.");
         }
 
-        AssetBus::MultiHandler::BusDisconnect();
```
Comment (jackalbe) [inline Code/Framework/AzCore/AzCore/Asset/AssetContainer.cpp:35]: should we even keep the AssetBus since it is possible to get it into a dead lock state?

**Review by jackalbe — APPROVED** [review]

**Review by AMZN-stankowi — APPROVED** [review]

File: Code/Framework/AzCore/AzCore/Asset/AssetContainer.cpp (hunk 1 above)
Comment (amzn-mike) [reply Code/Framework/AzCore/AzCore/Asset/AssetContainer.cpp:35]: I'm not sure I follow.  AssetBus should have been entirely removed from the AssetContainer.  It should only be using AssetLoadBus (the private ebus)

File: Code/Framework/AzCore/AzCore/Asset/AssetManager.cpp [hunk 2]
```diff
@@ -1857,6 +1860,7 @@ namespace AZ::Data
             AZStd::lock_guard<AZStd::recursive_mutex> assetLock(m_assetMutex);
             m_reloads.erase(asset.GetId());
         }
+        AssetLoadBus::Event(asset.GetId(), &AssetLoadBus::Events::OnAssetReady, asset); // Broadcast to any containers first
```
Comment (lumberyard-employee-dm) [inline Code/Framework/AzCore/AzCore/Asset/AssetManager.cpp:1863]: Shouldn't this be OnAssetReloadError?

Comment (amzn-mike) [reply Code/Framework/AzCore/AzCore/Asset/AssetManager.cpp]: Yep... fixed!

**Pull request merged**